File: services/ai/gateway-2.0/src/core/usage/tracker.ts
/**
 * Usage tracker – manages per-user rate-limiting with a sliding-window
 * approach backed by Redis, and logs usage events to PostgreSQL.
 *
 * Ported from the Go gateway `internal/usage` package.
 */

import type { FastifyBaseLogger } from "fastify";
import type { Redis } from "ioredis";
import type { GatewayConfig } from "../../config.js";
import type { PgPool } from "../../db/postgres.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface UsageInfo {
  tier: string;
  remaining: number;
  max: number;
  nextRechargeAt?: Date;
  fullRechargeAt?: Date;
}

export interface LogUsageParams {
  userId: string;
  channelType: string;
  tier: string;
  processingMs: number;
  model: string;
}

// ---------------------------------------------------------------------------
// Tracker
// ---------------------------------------------------------------------------

export class UsageTracker {
  private readonly config: GatewayConfig;
  private readonly redis: Redis;
  private readonly db: PgPool;
  private readonly logger: FastifyBaseLogger;

  constructor(
    config: GatewayConfig,
    redis: Redis,
    db: PgPool,
    logger: FastifyBaseLogger
  ) {
    this.config = config;
    this.redis = redis;
    this.db = db;
    this.logger = logger;
  }

  // -------------------------------------------------------------------------
  // Helpers
  // -------------------------------------------------------------------------

  private usageKey(userId: string): string {
    return `usage:${userId}:slots`;
  }

  // -------------------------------------------------------------------------
  // Check & Consume
  // -------------------------------------------------------------------------

  /**
   * Consume one usage slot for the given user.
   *
   * Returns `{ remaining }` where:
   *  - `remaining >= 0` – the number of slots still available *after* this
   *    consumption.
   *  - `remaining < 0`  – the user has exhausted their allowance; nothing was
   *    consumed.
   */
  async checkAndConsume(
    userId: string,
    _channelType: string
  ): Promise<{ remaining: number }> {
    try {
      const key = this.usageKey(userId);
      const rechargeDurationMs = this.config.freeRechargeHours * 3_600_000;
      const maxMessages = this.config.freeMaxMessages;
      const now = Date.now();

      // Retrieve all recorded timestamps from the Redis list.
      const timestamps = await this.redis.lrange(key, 0, -1);

      let activeCount = 0;
      for (const ts of timestamps) {
        const usedAtMs = parseInt(ts, 10) * 1_000; // stored as unix seconds
        if (now - usedAtMs < rechargeDurationMs) {
          activeCount++;
        }
      }

      const remaining = maxMessages - activeCount;
      if (remaining <= 0) {
        return { remaining: -1 };
      }

      // Record the current timestamp (unix seconds) at the head of the list.
      const nowUnix = Math.floor(now / 1_000);
      await this.redis.lpush(key, String(nowUnix));

      // Trim so we never store more entries than maxMessages.
      try {
        await this.redis.ltrim(key, 0, maxMessages - 1);
      } catch (err) {
        this.logger.warn({ err }, "Failed to trim usage list");
      }

      // Set a generous TTL so the key self-cleans.
      const maxTTLSeconds = maxMessages * this.config.freeRechargeHours * 3_600;
      try {
        await this.redis.expire(key, maxTTLSeconds);
      } catch (err) {
        this.logger.warn({ err }, "Failed to set usage TTL");
      }

      return { remaining: remaining - 1 };
    } catch (err) {
      this.logger.error({ err, userId }, "Failed in checkAndConsume");
      throw err;
    }
  }

  // -------------------------------------------------------------------------
  // Usage Info
  // -------------------------------------------------------------------------

  /**
   * Return the current usage information for a user without consuming a slot.
   */
  async getUsageInfo(userId: string): Promise<UsageInfo> {
    const key = this.usageKey(userId);
    const rechargeDurationMs = this.config.freeRechargeHours * 3_600_000;
    const maxMessages = this.config.freeMaxMessages;
    const now = Date.now();

    let timestamps: string[];
    try {
      timestamps = await this.redis.lrange(key, 0, -1);
    } catch (err) {
      this.logger.warn({ err, userId }, "Failed to read usage slots for info");
      return { tier: "free", remaining: maxMessages, max: maxMessages };
    }

    let earliestUsedMs = 0;
    let latestUsedMs = 0;
    let activeCount = 0;

    for (const ts of timestamps) {
      const usedAtMs = parseInt(ts, 10) * 1_000;
      if (now - usedAtMs < rechargeDurationMs) {
        activeCount++;
        if (earliestUsedMs === 0 || usedAtMs < earliestUsedMs) {
          earliestUsedMs = usedAtMs;
        }
        if (latestUsedMs === 0 || usedAtMs > latestUsedMs) {
          latestUsedMs = usedAtMs;
        }
      }
    }

    const info: UsageInfo = {
      tier: "free",
      remaining: maxMessages - activeCount,
      max: maxMessages,
    };

    if (activeCount > 0) {
      info.nextRechargeAt = new Date(earliestUsedMs + rechargeDurationMs);
      info.fullRechargeAt = new Date(latestUsedMs + rechargeDurationMs);
    }

    return info;
  }

  // -------------------------------------------------------------------------
  // Logging
  // -------------------------------------------------------------------------

  /**
   * Persist a usage event to the `gateway_usage_log` table.
   *
   * Failures are caught and logged – they must never break the request flow.
   */
  async logUsage(params: LogUsageParams): Promise<void> {
    try {
      await this.db.query(
        `INSERT INTO gateway_usage_log
           (user_id, channel_type, tier, processing_ms, model, created_at)
         VALUES ($1, $2, $3, $4, $5, NOW())`,
        [
          params.userId,
          params.channelType,
          params.tier,
          params.processingMs,
          params.model,
        ]
      );
    } catch (err) {
      this.logger.error(
        { err, userId: params.userId },
        "Failed to log usage event"
      );
    }
  }
}

File: services/ai/gateway/internal/usage/tracker.go
package usage

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/rs/zerolog"

	"github.com/stocktracker/gateway/internal/config"
	"github.com/stocktracker/gateway/internal/db"
)

// UsageInfo contains usage information for a user
type UsageInfo struct {
	Tier           string     `json:"tier"`
	Remaining      int        `json:"remaining"`
	Max            int        `json:"max"`
	NextRechargeAt *time.Time `json:"next_recharge_at,omitempty"`
	FullRechargeAt *time.Time `json:"full_recharge_at,omitempty"`
}

// Tracker manages message usage tracking and recharge
type Tracker struct {
	config   *config.Config
	redis    *db.RedisDB
	database *db.PostgresDB
	logger   zerolog.Logger
}

// NewTracker creates a new usage tracker
func NewTracker(cfg *config.Config, redis *db.RedisDB, database *db.PostgresDB, logger zerolog.Logger) *Tracker {
	return &Tracker{config: cfg, redis: redis, database: database, logger: logger}
}

func (t *Tracker) usageKey(userID string) string {
	return fmt.Sprintf("usage:%s:slots", userID)
}

// CheckAndConsume checks if the user has remaining messages and consumes one
// Returns remaining count (>= 0 means success, < 0 means exhausted)
func (t *Tracker) CheckAndConsume(ctx context.Context, userID, channelType string) (int, error) {
	key := t.usageKey(userID)
	rechargeDuration := time.Duration(t.config.FreeRechargeHours) * time.Hour
	maxMessages := t.config.FreeMaxMessages
	now := time.Now()

	// Get all timestamps from the list
	timestamps, err := t.redis.LRange(ctx, key, 0, -1)
	if err != nil {
		return 0, fmt.Errorf("failed to read usage slots: %w", err)
	}

	// Count active (non-recharged) slots
	activeCount := 0
	for _, ts := range timestamps {
		tsUnix, _ := strconv.ParseInt(ts, 10, 64)
		usedAt := time.Unix(tsUnix, 0)
		if now.Sub(usedAt) < rechargeDuration {
			activeCount++
		}
	}

	remaining := maxMessages - activeCount
	if remaining <= 0 {
		return -1, nil
	}

	// Consume a slot
	if err := t.redis.LPush(ctx, key, now.Unix()); err != nil {
		return 0, fmt.Errorf("failed to record usage: %w", err)
	}

	// Trim to max slots
	if err := t.redis.LTrim(ctx, key, 0, int64(maxMessages-1)); err != nil {
		t.logger.Warn().Err(err).Msg("Failed to trim usage list")
	}

	// Set TTL for auto-cleanup (all slots recharge after maxMessages * rechargeHours)
	maxTTL := time.Duration(maxMessages) * rechargeDuration
	if err := t.redis.Expire(ctx, key, maxTTL); err != nil {
		t.logger.Warn().Err(err).Msg("Failed to set usage TTL")
	}

	return remaining - 1, nil // -1 because we just consumed one
}

// GetUsageInfo returns usage information for a user
func (t *Tracker) GetUsageInfo(ctx context.Context, userID string) (*UsageInfo, error) {
	key := t.usageKey(userID)
	rechargeDuration := time.Duration(t.config.FreeRechargeHours) * time.Hour
	maxMessages := t.config.FreeMaxMessages
	now := time.Now()

	timestamps, err := t.redis.LRange(ctx, key, 0, -1)
	if err != nil {
		return &UsageInfo{Tier: "free", Remaining: maxMessages, Max: maxMessages}, nil
	}

	// Find active slots and earliest recharge
	var earliestUsed time.Time
	var latestUsed time.Time
	activeCount := 0

	for _, ts := range timestamps {
		tsUnix, _ := strconv.ParseInt(ts, 10, 64)
		usedAt := time.Unix(tsUnix, 0)
		if now.Sub(usedAt) < rechargeDuration {
			activeCount++
			if earliestUsed.IsZero() || usedAt.Before(earliestUsed) {
				earliestUsed = usedAt
			}
			if latestUsed.IsZero() || usedAt.After(latestUsed) {
				latestUsed = usedAt
			}
		}
	}

	info := &UsageInfo{
		Tier:      "free",
		Remaining: maxMessages - activeCount,
		Max:       maxMessages,
	}

	if activeCount > 0 {
		nextRecharge := earliestUsed.Add(rechargeDuration)
		info.NextRechargeAt = &nextRecharge
		fullRecharge := latestUsed.Add(rechargeDuration)
		info.FullRechargeAt = &fullRecharge
	}

	return info, nil
}
